src/client/AppClientUIHost.ts

```diff
@@ -0,0 +1,69 @@
+import {
+    ClientEmbeddedSDKActions,
+    IClientEmbeddedSDKResonse,
+    IClientRoomInfo,
+    IClientUserInfo,
+} from './definition';
+
+/**
+ * Represents the host which handlers API calls from external components.
+ */
+export abstract class AppClientUIHost {
+    /**
+     * The message emitter who calling the API.
+     */
+    private emitter!: MessageEventSource;
+    constructor() {
+        this.initialize();
+    }
+    /**
+     * initialize the AppClientUIHost by registering window `message` listener
+     */
+    public initialize() {
+        window.addEventListener('message', async ({ data, source }) => {
+            this.emitter = source;
+
+            if (!data.hasOwnProperty('rcEmbeddedSDK')) {
+                return;
+            }
+            const { rcEmbeddedSDK: { action, id } } = data;
+
+            switch (action) {
+                case ClientEmbeddedSDKActions.GET_USER_INFO:
+                    this.handleAction(action, id, await this.getClientUserInfo());
+                case ClientEmbeddedSDKActions.GET_ROOM_INFO:
+                    this.handleAction(action, id, await this.getClientRoomInfo());
+            }
+        });
+    }
+    /**
+     * Get the current user's information.
+     */
+    public abstract async getClientUserInfo(): Promise<IClientUserInfo>;
+    /**
+     * Get the opened room's information.
+     */
+    public abstract async getClientRoomInfo(): Promise<IClientRoomInfo>;
+    /**
+     * Handle the action sent from the external component.
+     * @param action the name of the action
+     * @param id the unique id of the  API call
```

Comment:
What is the `id`? I'm not sure I understand this part of it..

Reply:
https://github.com/RocketChat/Rocket.Chat.Apps-engine/blob/50a4c9560842e4df2928785cdb10d30a5dbfe61c/src/client/AppClientEmbeddedSDK.ts#L56-L64

In the AppClientEmbeddedSDK, we call API by specifying an action, which contains a unique key called `id`. After receiving callbacks from the host, we will delete the corresponding action by specifying the `id` from the callbacks map to indicate this action has been resolved. as shown below:
https://github.com/RocketChat/Rocket.Chat.Apps-engine/blob/50a4c9560842e4df2928785cdb10d30a5dbfe61c/src/client/AppClientEmbeddedSDK.ts#L44-L51